### ISSUE TITLE
Nested Forms validation

### ISSUE THREAD

**Issue by aleshchenko** [angular/angular.js#5858]
It would be great to add some optional attribute to form element or div with ng-form attribute which will say that validation errors of current form shouldn't affect validation of parent form.
Here is what I mean:

```
<div ng-form="parentForm">
  <input...>some valid input</input>
  <div ng-form="childForm" omit-parent-validation="true">
    <input...>some INVALID input</input>
  </div>
</div>
```

Result:
parentForm.$valid -> true, childForm.$valid -> false
Same logic may be applied for
parentForm.$pristine -> true, childForm.$pristine -> false

**Comment by alessiodm** [comment]
I wanted to reach a slightly different result from the one described by this issue: i.e., I just wanted to detach some controls from the containing form. This directive does the job, simply removing the `$formController` from the `inheritedData` chain and causing the `ngModelController`s to register a `nullFormCtrl` for their `parentForm`:
```javascript
.directive('teEvictForm', function() {
    return {
        restrict : 'A',
        link: {
            pre: function (scope, iElement) {
                iElement.data('$formController', null);
            }
        }
    };
})
```
I think it might apply to subforms as well, so I'm just posting it here.

**Comment by iyel** [comment]
@alessiodm your solution breaks form from updating itself

**Comment by petebacondarwin** [comment]
There is a PR for this at #10193

**Comment by williamweckl** [comment]
+ 1
Using @91K00 solution.

**Comment by evilaliv3** [comment]
+1

Using @91K00 solution also!

**Comment by rSafari** [comment]
First of all, thank you to @91K00 and @gonzaloruizdevilla for trying to fix the problem.

lets assume this structure:

```
<ng-form name="X1" novalidate>

    <ng-form name="X2" novalidate isolate-form>

        <input name="Input01" required />
        <p ng-show="X2.Input01.$touched && X2.Input01.$invalid">input is not valid</p>

        <input name="Input02" required />

        <input type="button" id="ButtonX2" value="Submit Nested Form" ng-disabled="X2.$invalid" />

    </ng-form>

<input name="Input03" required ng-minlength="5" />

<input type="button" id="ButtonX1" value="Submit Nested Form" ng-disabled="X1.$invalid" />

</ng-form> 
```

neither the ngFormTopLevel or @91K00 solution can handle this.

**tl;dr :**
ButtonX1 is dependent to nested form validation and it shouldn't !

**Test case 1:**
Step 1: Fill input3 with any text and more than 5 character.
Expected: ButtonX1 should be enable.
Result: ButtonX1 still disabled.

**Test case 2:**
Step 1: Fill input1 with any text.
Step 2: Fill input2 with any text.
Expected: ButtonX2 should be enable.
Result: ButtonX2 is enabled.

**Test case 3:**
Step 1: Fill input3 with any text and more than 5 character.
Step 2: Fill input1 with any text.
Step 2: Fill input2 with any text.
Expected: ButtonX1 and ButtonX2 should be enable.
Result: ButtonX1 and ButtonX2 is enabled.

and the other problem is the P tag inside the nested form does not show when the Input01 is invalid.
i tried both the isolateForm and the ngFormTopLevel but both of them have this problem.

**Comment by 91K00** [comment]
Hi rSafari,
If you add "ng-model" attribute on each input text, it works.
For the message inside the P, I don't know why the $touched control is not working... You can use $dirty or ng-message module instead.

**Comment by rSafari** [comment]
Yhank you @91K00 for your answer.

of course i have ng-model for my inputs. i just forgot to add it in this structure ... ( i updated my post)

the main problem is that i have to validate the nested form before i can submit the parent form. it means i cant submit the parent form until i add something valid to the nested form inputs.

**Comment by rSafari** [comment]
stackoverflow related topic :
http://stackoverflow.com/questions/37938530/validate-nested-forms-without-affecting-the-parent-form

**Comment by penfold** [comment]
This solution no longer works (tested in v1.6.2) due to an exception when copying the original form controller.

> Error: [ng:cpws] Can't copy! Making copies of Window or Scope instances is not supported.

Any suggestions for a workaround?

**Comment by Erffun** [comment]
also not working on v1.6.3

**Comment by Narretz** [comment]
why do need to copy the form Controller anyway? It should be enough to simply call parent.$removeControl(self) in the directive. If the form does not have a parent form controller, it doesn't propagate change upwards.

**Comment by momega** [comment]
Here is working code for angular >=1.6.2, based recent comment. https://github.com/momega/isolate-form

**Comment by Narretz** [comment]
Isolate forms aren't in scope for core because the logic required is too heavy. See https://github.com/angular/angular.js/pull/10193#discussion_r66080985 You can detach form controls, however.
